File: java-workshop/part-4-behavior/src/test/java/helper/InMemoryPaymentGateway.java
package helper;

import domain.PaymentRequest;
import domain.PaymentResult;
import domain.PaymentStatus;
import gateway.PaymentGateway;

import java.util.ArrayList;
import java.util.List;

public class InMemoryPaymentGateway implements PaymentGateway {

    private boolean willDecline = false;
    private PaymentResult lastResult;
    private final List<String> refunds = new ArrayList<>();

    public void willDecline() {
        this.willDecline = true;
    }

    @Override
    public PaymentResult process(PaymentRequest request) {
        if (willDecline) {
            lastResult = new PaymentResult(request.orderId(), PaymentStatus.FAILED, null);
        } else {
            lastResult = new PaymentResult(request.orderId(), PaymentStatus.SUCCESS, "TXN-" + request.orderId());
        }
        return lastResult;
    }

    @Override
    public boolean refund(String transactionId) {
        refunds.add(transactionId);
        return true;
    }

    public boolean wasCharged() {
        return lastResult != null && lastResult.status() == PaymentStatus.SUCCESS;
    }

    public String lastTransactionId() {
        return lastResult != null ? lastResult.transactionId() : null;
    }

    public boolean wasRefunded(String transactionId) {
        return refunds.contains(transactionId);
    }
}

File: csharp-workshop/Part4.Behavior.Tests/Helper/InMemoryPaymentGateway.cs
using Part4.Behavior.Domain;
using Part4.Behavior.Gateway;

namespace Part4.Behavior.Tests.Helper;

public class InMemoryPaymentGateway : IPaymentGateway
{
    private bool _willDecline = false;
    private PaymentResult? _lastResult;
    private readonly List<string> _refunds = new();

    public void WillDecline()
    {
        _willDecline = true;
    }

    public PaymentResult Process(PaymentRequest request)
    {
        if (_willDecline)
        {
            _lastResult = new PaymentResult(request.OrderId, PaymentStatus.Failed, null);
        }
        else
        {
            _lastResult = new PaymentResult(request.OrderId, PaymentStatus.Success, "TXN-" + request.OrderId);
        }
        return _lastResult;
    }

    public bool Refund(string transactionId)
    {
        _refunds.Add(transactionId);
        return true;
    }

    public bool WasCharged() => _lastResult != null && _lastResult.Status == PaymentStatus.Success;

    public string? LastTransactionId() => _lastResult?.TransactionId;

    public bool WasRefunded(string transactionId) => _refunds.Contains(transactionId);
}
